algebra/src/fields/mod.rs

```diff
@@ -64,37 +65,27 @@ pub trait Field:
     + Sync
     + 'static
     + Eq
+    + One
     + Ord
     + Neg<Output = Self>
     + UniformRand
+    + Zero
     + Sized
     + Hash
     + From<u128>
     + From<u64>
     + From<u32>
     + From<u16>
     + From<u8>
-    + for<'a> Add<&'a Self, Output = Self>
```

Comment:
Do we need to remove these bounds?

Reply:
In particular, the following error in https://github.com/scipr-lab/zexe/pull/54#issuecomment-574195828 could be because these bounds were removed, and so `rustc` errored out.
```rust
error[E0308]: mismatched types                                                                                                                                                                                                                                                   
  --> algebra/src/curves/models/short_weierstrass_jacobian.rs:81:44                                                                                                                                                                                                              
   |                                                                                                                                                                                                                                                                             
81 |         let x3b = P::add_b(&((x.square() * &x) + &P::mul_by_a(&x)));                                                                                                                                                                                                        
   |                                            ^^                                                                                                                                                                                                                               
   |                                            |                                                                                                                                                                                                                                
   |                                            expected associated type, found reference                                                                                                                                                                                        
   |                                            help: consider removing the borrow: `x`                                                                                                                                                                                          
   |                                                                                                                                                                                                                                                                             
   = note: expected associated type `<P as curves::models::ModelParameters>::BaseField`                                                                                                                                                                                          
                    found reference `&<P as curves::models::ModelParameters>::BaseField`  
```

Reply:
Please consider https://gist.github.com/1e4f5aaf15f910a53896f4dfb9f68627 which is a compilation log on my branch [num_traits_2](https://github.com/huitseeker/zexe/tree/num_traits_2) at [723b6a126](https://github.com/huitseeker/zexe/commit/723b6a126d5dd8ca7ba6a312ec955ea0124bc5aa). As we discussed, this branch does not contain any of the reference adjustments, does contain the traits bounds above, along with everything else on this PR.

(But fair point on the removed trait bounds, I've reintroduced them to the present PR).

Reply:
Hmm I was able to get it to compile with the following patch:

```
From e1c777280d6956c4dbf9015d34072e85c762ac80 Mon Sep 17 00:00:00 2001
From: Pratyush Mishra <pratyushmishra@berkeley.edu>
Date: Wed, 15 Jan 2020 08:06:21 -0800
Subject: [PATCH] Compile with refs

---
 algebra/src/curves/models/short_weierstrass_jacobian.rs   | 4 ++--
 algebra/src/curves/models/short_weierstrass_projective.rs | 4 ++--
 algebra/src/fields/mod.rs                                 | 4 +---
 3 files changed, 5 insertions(+), 7 deletions(-)

diff --git a/algebra/src/curves/models/short_weierstrass_jacobian.rs b/algebra/src/curves/models/short_weierstrass_jacobian.rs
index be4924a..318c33e 100644
--- a/algebra/src/curves/models/short_weierstrass_jacobian.rs
+++ b/algebra/src/curves/models/short_weierstrass_jacobian.rs
@@ -129,8 +129,8 @@ impl<P: Parameters> Add<Self> for GroupAffine<P> {
 impl<'a, P: Parameters> AddAssign<&'a Self> for GroupAffine<P> {
     fn add_assign(&mut self, other: &'a Self) {
         let lambda = (other.y - &self.y) / &(other.x - &self.y);
-        let x3 = lambda * lambda - &self.x - &other.x;
-        let y3 = (self.x - &x3) * lambda - &self.y;
+        let x3 = lambda * &lambda - &self.x - &other.x;
+        let y3 = (self.x - &x3) * &lambda - &self.y;
 
         self.x = x3;
         self.y = y3;
diff --git a/algebra/src/curves/models/short_weierstrass_projective.rs b/algebra/src/curves/models/short_weierstrass_projective.rs
index 1602a35..1ab87cc 100644
--- a/algebra/src/curves/models/short_weierstrass_projective.rs
+++ b/algebra/src/curves/models/short_weierstrass_projective.rs
@@ -132,8 +132,8 @@ impl<P: Parameters> Add<Self> for GroupAffine<P> {
 impl<'a, P: Parameters> AddAssign<&'a Self> for GroupAffine<P> {
     fn add_assign(&mut self, other: &'a Self) {
         let lambda = (other.y - &self.y) / &(other.x - &self.y);
-        let x3 = lambda * lambda - &self.x - &other.x;
-        let y3 = (self.x - &x3) * lambda - &self.y;
+        let x3 = lambda * &lambda - &self.x - &other.x;
+        let y3 = (self.x - &x3) * &lambda - &self.y;
 
         self.x = x3;
         self.y = y3;
diff --git a/algebra/src/fields/mod.rs b/algebra/src/fields/mod.rs
index 37d0391..a49c322 100644
--- a/algebra/src/fields/mod.rs
+++ b/algebra/src/fields/mod.rs
@@ -77,11 +77,9 @@ pub trait Field:
     + From<u32>
     + From<u16>
     + From<u8>
-    + for<'a> Add<&'a Self, Output = Self>
     + Add<Self, Output = Self>
     + for<'a> Add<&'a Self, Output = Self>
     + for<'a> Sub<&'a Self, Output = Self>
-    + for<'a> Mul<&'a Self, Output = Self>
     + Mul<Self, Output = Self>
     + for<'a> Mul<&'a Self, Output = Self>
     + for<'a> Div<&'a Self, Output = Self>
@@ -341,7 +339,7 @@ pub fn batch_inversion<F: Field>(v: &mut [F]) {
         .zip(prod.into_iter().rev().skip(1).chain(Some(F::one())))
     {
         // tmp := tmp * g.z; g.z := tmp * s = 1/z
-        let newtmp = tmp * &f;
+        let newtmp = tmp * *f;
         *f = tmp * &s;
         tmp = newtmp;
     }
-- 
2.17.1
```

Reply:
This takes a tad more work than your patch (which does the work for the algebra module), but happy to take things in this direction. Stand by.

Reply:
Awesome, thanks so much for your patience =)